src/event_pusher/mod_event_pusher_sns.erl

```diff
@@ -0,0 +1,240 @@
+
+    WorkerNum = gen_mod:get_opt(pool_size, Opts, 100),
+    wpool:start_sup_pool(pool_name(Host), [{workers, WorkerNum}]),
+
+    %% Check for required options
+    RequiredOptions = [access_key_id, secret_access_key, region, account_id, sns_host],
+    lists:foreach(
+        fun(Option) ->
+            case opt(Host, Option) of
+                undefined ->
+                    error({missing_required_option, Option});
+                _ -> ok
+            end
+         end, RequiredOptions),
+
+    ok.
+
+-spec stop(Host :: ejabberd:server()) -> ok.
+stop(Host) ->
+    wpool:stop(pool_name(Host)),
+    ok.
+
+push_event(_, #user_status_event{jid = UserJID, status = Status}) ->
+    user_presence_changed(UserJID, Status == online);
+push_event(_, #chat_event{direction = in, from = From, to = To, packet = Packet}) ->
+    handle_packet(From, To, Packet);
+push_event(_, _) ->
+    ok.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+-spec user_presence_changed(UserJID :: ejabberd:jid(), IsOnline :: boolean()) -> ok.
+user_presence_changed(#jid{lserver = Host} = UserJID, IsOnline) ->
+    Topic = opt(Host, presence_updates_topic),
+    case Topic of
+        undefined ->
+            skip;
+        Topic ->
+            UserGUID = user_guid(Host, UserJID),
+            Content = #{user_id => UserGUID, present => IsOnline},
+            TopicARN = make_topic_arn(Host, Topic),
+            Attributes = message_attributes(Host, TopicARN, UserJID, IsOnline),
+            async_publish(Host, TopicARN, Content, Attributes)
+    end,
+    ok.
+
+%% @doc Handles packet and if needed publishes SNS notification
+-spec handle_packet(From :: ejabberd:jid(), To :: ejabberd:jid(),
+                    Packet :: jlib:xmlel()) -> ok | skip.
+handle_packet(From = #jid{lserver = Host}, To, Packet) ->
+    ?DEBUG("SNS Packet handle~n    from ~p ~n    to ~p~n    packet ~p.", [From, To, Packet]),
+
+    case {get_topic(Host, Packet), exml_query:subelement(Packet, <<"body">>)} of
+        {undefined, _} -> %% Skip if there is no topic set in configuration for the packet type
+            skip;
+        {_, undefined} -> %% Skip if there is no message body in the packet
+            skip;
+        {Topic, BodyTag} ->
+            FromGUID = user_guid(Host, From),
+            ToGUID = user_guid(Host, To),
+            MessageBody = exml_query:cdata(BodyTag),
+            Content = #{from_user_id => FromGUID,
+                        to_user_id => ToGUID,
+                        message => MessageBody},
+
+            TopicARN = make_topic_arn(Host, Topic),
+            Attributes = message_attributes(Host, TopicARN, From, To, message_type(Packet), Packet),
+
+            async_publish(Host, TopicARN, Content, Attributes)
+    end.
+
+%% @doc Start publish process notification to AWS SNS service. Content should be valid JSON term
```

Comment:
Start publish process notification to AWS SNS service. Content should be a valid JSON term